EIPS/eip-7701.md

```diff
@@ -0,0 +1,213 @@
+---
+eip: 7701
+title: Native Account Abstraction with EOF
+description: A variant of RIP-7560 transactions relying on EOF Smart Contract Accounts
+author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Alex Forshtat (@forshtat), Dror Tirosh (@drortirosh), Shahaf Nacson (@shahafn)
+discussions-to: https://ethereum-magicians.org/t/eip-7701-native-account-abstraction-with-eof/19893
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-05-01
+requires: 3540
+---
+
+## Abstract
+
+This proposal describes a slight variation of the Native Account Abstraction design fully described in RIP-7560.
+This version's difference compared to the original proposal is in relying on features of "EVM Object Format" to
+distinguish between validation and execution code sections.
+
+## Motivation
+
+Talking about Full Native Account Abstraction, the fundamental idea any solution has to address is a
+mechanism for a Smart Contract Account to separate its validation and execution code sections.
+
+RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
+on using some higher-level abstraction.
+
+In its current form, RIP-7560 transactions use Solidity method selectors in order to achieve this separation.
+
+This, however, is far from ideal as this approach "leaks" the concept from a programming language widely used
+in the EVM into the core design of the Ethereum protocol.
+
+While there is no purely technical reason to disallow it and there are already instances of Solidity code
+being "enshrined" in the Ethereum protocol, e.g. the validator deposit contract, such violation of abstraction levels
+often lead to unnecessary technical debt and are worth avoiding if possible.
+
+Additionally, method selectors provide very weak indication of the contract developer's decision to become a
+participant in Native Account Abstraction transaction flow.
+The chance of accidentally exposing a function with a colliding 4 byte
+method identifier that returns a valid 4 byte magic to indicate approval is pretty low, but a malicious developer
+can easily hide such a function giving it an innocent name, making it hard to spot a Native Account Abstraction entity.
+
+This issue to some extent is also present in [ERC-4337](./eip-4337.md).
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with [EIP-3540](./eip-3540),
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### System-level code entry points
+
+Modify the EOF container format to consist of the following sections:
+
+```
+container := header, body
+header :=
+    magic, version,
+    kind_types, types_size,
+    kind_entrypoints, entrypoints_size,
```

Comment:
kind_entrypoints line should be optional, so EOF1 will resemble EOF2-for-non-AA
(being non-optional, it requires EOF2 to have entrypoints_size=0).
since each line starts with "kind_*" prefix, there is no overhead to make a line optional.
Also,  it better move the kind_entrypoints AFTER the kind_code, near the optional kind_containers